### PR TITLE
Solutions for Chapter 4

### PR DESCRIPTION
**Pull request by rahulbhargav**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +486,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap e) _ = Trap e
+    (<*>) (Reward f) s = fmap f s
```
Comment (vrom911) [inline src/Chapter4.hs:493]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -611,6 +638,10 @@ Implement the 'Monad' instance for our lists.
   maybe a few) to flatten lists of lists to a single list.
 -}
 
+instance Monad List where
+  (>>=) :: List a -> (a -> List b) -> List b
+  (>>=) Empty _ = Empty
+  (>>=) (Cons x xs) f = concatList (f x) (xs >>= f)
```
Comment (vrom911) [inline src/Chapter4.hs:644]: you can also just use `fmap` in here. As `fmap` already handles empty lists, you don't need to pattern match:

```haskell
xs >>= f = concatList (fmap f xs)
```

File: src/Chapter4.hs [hunk 3]
```diff
@@ -629,7 +660,8 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+-- andM m1 m2 = pure (&&) <*> m1 <*> m2 -- (Without >>= and anonymous function)
```
Comment (vrom911) [inline src/Chapter4.hs:663]: The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face: 


If you don't use the `>>=` operator, then the `Monad` constraint is too strong, and it is enough to specify only `Applicative`. However, with `Applicative` you are not able to implement a lazier version of `andM`.

File: src/Chapter4.hs [hunk 4]
```diff
@@ -673,6 +705,22 @@ Specifically,
...
+  | Node e (BTree e, BTree e)
+
+instance Functor BTree where
+  fmap :: (a -> b) -> BTree a -> BTree b
+  fmap _ Leaf = Leaf
+  fmap f (Node e (l,r)) = Node (f e) ((fmap f l), (fmap f r))
+
+reverseTree :: BTree e -> BTree e
+reverseTree Leaf = Leaf
+reverseTree (Node e (l, r)) = Node e ((reverseTree l), (reverseTree r))
```
Comment (vrom911) [inline src/Chapter4.hs]: For reverse, you actually need to swap subtrees. This function is not changing the Tree, as you here would return the same that you get 🌴

**Review by vrom911 — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun 🌮 🥳 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger 🤗

File: src/Chapter4.hs (hunk 4 above)
Comment (rahulbhargav) [reply src/Chapter4.hs]: Oops. Thanks for pointing it out. Fixed it now 😊

File: src/Chapter4.hs (hunk 2 above)
Comment (rahulbhargav) [reply src/Chapter4.hs:644]: Correct. I didn't think of this. Thank you 😊

Comment (rahulbhargav) [reply src/Chapter4.hs:644]: But here, concatList accepts two lists as input. I guess what you're suggesting here is to use a `flatten` method instead?

File: src/Chapter4.hs (hunk 3 above)
Comment (rahulbhargav) [reply src/Chapter4.hs:663]: Understood. Thanks for a clear explanation 😊

File: src/Chapter4.hs (hunk 2 above)
Comment (vrom911) [reply src/Chapter4.hs:644]: Oh, yes, sorry, in the standard library, `concat` is for list flattering 😄

**Pull request merged**